safekeeper/spec/ProposerAcceptorStatic.tla

```diff
@@ -0,0 +1,448 @@
+        \* done (if it is not max_vote_acc).
+       /\ acc_state[max_vote_acc].term = prop_state[p].term
+       /\ prop_state' = [prop_state EXCEPT ![p].state = "leader",
+                                           ![p].termHistory = prop_th,
+                                           ![p].wal = acc_state[max_vote_acc].wal
+                        ]
+       /\ elected_history' = Upsert(elected_history, prop_state[p].term, 1, LAMBDA c: c + 1)
+       /\ UNCHANGED <<acc_state, committed>>
+
+
+\* Acceptor a learns about elected proposer p's term. In impl it matches to
+\* VoteRequest/VoteResponse exchange when leader is already elected and is not
+\* interested in the vote result.
+UpdateTerm(p, a) ==
+  /\ prop_state[p].state = "leader"
+  /\ acc_state[a].term < prop_state[p].term
+  /\ acc_state' = [acc_state EXCEPT ![a].term = prop_state[p].term]
+  /\ UNCHANGED <<prop_state, committed, elected_history>>
+
+\* Find highest common point (LSN of the first divergent record) in the logs of
+\* proposer p and acceptor a. Returns <term, lsn> of the highest common point.
+FindHighestCommonPoint(prop_th, acc_th, acc_flush_lsn) ==
+    LET
+      \* First find index of the highest common term.
+      \* It must exist because we initialize th with <0, 1>.
+      last_common_idx == Maximum({i \in 1..Min(Len(prop_th), Len(acc_th)): prop_th[i].term = acc_th[i].term})
+      last_common_term == prop_th[last_common_idx].term
+      \* Now find where it ends at both prop and acc and take min. End of term
+      \* is the start of the next unless it is the last one; there it is
+      \* flush_lsn in case of acceptor. In case of proposer it is the current
+      \* writing position, but it can't be less than flush_lsn, so we
+      \* take flush_lsn.
+      acc_common_term_end == IF last_common_idx = Len(acc_th) THEN acc_flush_lsn ELSE acc_th[last_common_idx + 1].lsn
+      prop_common_term_end == IF last_common_idx = Len(prop_th) THEN acc_flush_lsn ELSE prop_th[last_common_idx + 1].lsn
+    IN
+      [term |-> last_common_term, lsn |-> Min(acc_common_term_end, prop_common_term_end)]
+
+\* Elected proposer p immediately truncates WAL (and term history) of acceptor a
+\* before starting streaming. Establishes nextSendLsn for a.
+\*
+\* In impl this happens at each reconnection, here we also allow to do it multiple times.
+TruncateWal(p, a) ==
+  /\ prop_state[p].state = "leader"
+  /\ acc_state[a].term = prop_state[p].term
+  /\ LET
+       hcp == FindHighestCommonPoint(prop_state[p].termHistory, AcceptorTermHistory(a), FlushLsn(a))
+       next_send_lsn == (a :> hcp.lsn) @@ prop_state[p].nextSendLsn
+     IN
+          \* Acceptor persists full history immediately; reads adjust it to the
+          \* really existing wal with AcceptorTermHistory.
+       /\ acc_state' = [acc_state EXCEPT ![a].termHistory = prop_state[p].termHistory,
+                                         \* note: SubSeq is inclusive, hence -1.
```

Comment:
In the implementation, do we actually transmit the full term history?
Isn't that a long-term scalability issue because termHistory grows forever?

Reply:
We do. Right, it is. We have an issue for that and should fix it at some point. It is not that bad because even 65k terms take only ~1MB.